activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -272,7 +272,11 @@ def construct(ar_parent, parent, row, rs, seen, model_cache, aliases)
             construct(model, node, row, rs, seen, model_cache, aliases)
           else
             model = construct_model(ar_parent, node, row, model_cache, id, aliases)
-            model.readonly!
+            join_scope = node.reflection.klass.unscoped
+            join_scope = join_scope.instance_exec(nil, &node.reflection.scope) if node.reflection.scope
+            if join_scope.values[:readonly]
+              model.readonly!
+            end
```

Comment:
This implementation is reaching into a lot of internals of other objects. Can you refactor this to something a bit more contained?